### PR TITLE
Add update row to low level python

### PR DESCRIPTION
**Pull request by benjeffery**
Initial low level work for #1545 
WIP - only individuals done

### CODE REVIEW

**Comment by benjeffery** [comment]
@jeromekelleher Can I get a quick review here before the copy-paste-athon?

File: python/_tskitmodule.c [hunk 1]
```diff
@@ -779,6 +779,31 @@ uint32_converter(PyObject *py_obj, uint32_t *uint_out)
...
+    int ret = 0;
+
+    if (!PyArg_Parse(py_obj, "L", &temp_long)) {
+        goto out;
+    }
+    if (temp_long > TSK_MAX_ID) {
+        PyErr_SetString(PyExc_OverflowError, "Value too large for tskit id type");
+        goto out;
+    }
+    if (temp_long < TSK_NULL) {
```
Comment (jeromekelleher) [inline python/_tskitmodule.c:795]: Not sure this is worth checking at this level, but I don't feel strongly about it.

File: python/_tskitmodule.c [hunk 2]
```diff
@@ -865,14 +890,17 @@ IndividualTable_init(IndividualTable *self, PyObject *args, PyObject *kwds)
 }
 
 static PyObject *
-IndividualTable_add_row(IndividualTable *self, PyObject *args, PyObject *kwds)
+IndividualTable_add_or_update_row(IndividualTable *self, PyObject *args, PyObject *kwds)
```
Comment (jeromekelleher) [inline python/_tskitmodule.c:893]: I don't like this pattern - trying to back-infer how a method was called always seems like a bad idea, and leads to convoluted logic.

**Review by jeromekelleher — COMMENTED** [review]
Hmm, not sure about this pattern. I see what you're trying to do though, and avoiding duplicating all the arg parsing logic is definition worth doing.

Let me have a think about it and see if I can come up with anything better. So, hold off on the copy-paste-athon for a bit if you don't mind, please.

**Comment by jeromekelleher** [comment]
I've had a look at this, and actually the IndividualTable is the wrong place to start as it's by far and away the most complicated example. For the other tables, it'll be much simpler to just copy and paste and tweak in the obvious way than to try and factor out the common code. Try seeing what it looks like for EdgeTable, and then consider which option (i.e., trying to factor out the argument parsing code, or just copy-paste-altering) would be easier for a new contributor to understand.

File: python/_tskitmodule.c (hunk 1 above)
Comment (benjeffery) [reply python/_tskitmodule.c:795]: I've done this this way as it should cope with both 32 and 64 bit ids should we wish to support that in future.

File: python/_tskitmodule.c [hunk 3]
```diff
@@ -1981,6 +2125,45 @@ EdgeTable_add_row(EdgeTable *self, PyObject *args, PyObject *kwds)
...
 
+static PyObject *
+EdgeTable_update_row(EdgeTable *self, PyObject *args, PyObject *kwds)
+{
+    PyObject *ret = NULL;
+    int err;
+    tsk_id_t row_index = -1;
+    double left = 0.0;
+    double right = 1.0;
+    int parent;
```
Comment (jeromekelleher) [inline python/_tskitmodule.c:2136]: might as well save a big of vspace here, ``int parent, child;``. Same for left and right, we're not using the values we're setting, so no point in setting them.

File: python/_tskitmodule.c [hunk 4]
```diff
@@ -1981,6 +2125,45 @@ EdgeTable_add_row(EdgeTable *self, PyObject *args, PyObject *kwds)
...
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O&ddii|O", kwlist, &tsk_id_converter,
+            &row_index, &left, &right, &parent, &child, &py_metadata)) {
+        goto out;
+    }
+    if (py_metadata != Py_None) {
+        if (PyBytes_AsStringAndSize(py_metadata, &metadata, &metadata_length) < 0) {
+            goto out;
+        }
+    }
+    err = tsk_edge_table_update_row(
```
Comment (jeromekelleher) [inline python/_tskitmodule.c:2156]: Should probably cast the ints to ``tsk_id_t``, and metadata_length to ``tsk_size_t``. (No need to go over old code, but for new code good to do it more correctly)

**Review by jeromekelleher — COMMENTED** [review]
LGTM, minor points on type safety

File: python/_tskitmodule.c (hunk 4 above)
Comment (jeromekelleher) [reply python/_tskitmodule.c:2156]: Or, maybe we should just use your nice ``tsk_id_converter`` on all the tsk_id_t values, and save this annoying little dance?

File: python/_tskitmodule.c (hunk 3 above)
Comment (benjeffery) [reply python/_tskitmodule.c]: Yep, done in a couple of the methods

File: python/_tskitmodule.c (hunk 4 above)
Comment (benjeffery) [reply python/_tskitmodule.c]: I've used `tsk_id_converter` and cast the string lengths where needed (including ancestral_state etc.)

Comment (benjeffery) [reply python/_tskitmodule.c]: I've used `tsk_id_converter` and cast the string lengths where needed (including ancestral_state etc.)

**Review by jeromekelleher — COMMENTED** [review]
Looks good, thanks @benjeffery. Unfortunately I think we do need to get a bit more test coverage on the argument parsing error and default handling - perhaps we could add a handful of update_row invocations to where we're doing this stuff for add_row? Exhastive testing of all the options would be really tedious and a bit pointless, but we want some coverage on the error cases for the new methods (I see we do have good tests for the row_id case, all right though, which does give a lot of assurance that the error handling code is working properly).

I'm paranoid about this error handling stuff as it has turned into a real time-sink in the past trying to track down segfaults caused by weird behaviour in small mistakes with ArgParseTuple

**Comment by benjeffery** [comment]
>  perhaps we could add a handful of update_row invocations to where we're doing this stuff for add_row?

In that case I'm tempted to update the old `add_row` code so that the tests can be identical.

**Comment by jeromekelleher** [comment]
Changes look good, great to cut down on our technical dept a bit like this.

**Comment by benjeffery** [comment]
Ok, I think this is good to go. I didn't add the update tests where we are doing add row tests as they are in the higher-level tests. I also spotted some missing `add_row` tests.

**Review by jeromekelleher — APPROVED** [review]
Looks great, thanks @benjeffery!

**Pull request merged**